pkg/networkservice/common/connect/server.go

```diff
@@ -94,60 +125,67 @@ func (s *connectServer) Close(ctx context.Context, conn *networkservice.Connecti
 	return &empty.Empty{}, err
 }
 
-func (s *connectServer) client(ctx context.Context, conn *networkservice.Connection) networkservice.NetworkServiceClient {
+func (s *connectServer) client(ctx context.Context, conn *networkservice.Connection) *clientInfo {
 	logger := log.FromContext(ctx).WithField("connectServer", "client")
-
 	clientURL := clienturlctx.ClientURL(ctx)
-	if clientURL == nil {
-		err := errors.Errorf("clientURL not found for incoming connection: %+v", conn)
-		return injecterror.NewClient(err)
-	}
 
 	// First check if we have already requested some clientURL with this conn.GetID().
 	if connInfo, ok := s.connInfos.Load(conn.GetId()); ok {
 		if *connInfo.clientURL == *clientURL {
 			return connInfo.client
 		}
-		// For some reason we have changed the clientURL, so we need to close the existing client.
-		if _, clientErr := connInfo.client.Close(ctx, conn); clientErr != nil {
+
+		// For some reason we have changed the clientURL, so we need to close and delete the existing client.
+		if _, clientErr := connInfo.client.client.Close(ctx, conn); clientErr != nil {
 			logger.Warnf("failed to close client: %+v", clientErr)
 		}
-	}
 
-	// Fast path if we already have client for the clientURL, use it.
-	client, loaded := s.clients.Load(clientURL.String())
-	if !loaded {
-		// If not, create and LoadOrStore a new one.
-		newClient, cancel := s.newClient(clientURL)
-		client, loaded = s.clients.LoadOrStore(clientURL.String(), newClient)
-		if loaded {
-			// No one will use `newClient`, it should be canceled.
-			cancel()
-		}
+		s.closeClient(connInfo.client, connInfo.clientURL.String())
 	}
 
-	s.connInfos.Store(conn.GetId(), connectionInfo{
-		clientURL: clientURL,
-		client:    client,
+	var client *clientInfo
+	<-s.executor.AsyncExec(clientURL.String(), func() {
+		// Fast path if we already have client for the clientURL and we chould not reconnect, use it.
+		var loaded bool
+		client, loaded = s.clients.Load(clientURL.String())
+		if !loaded {
+			// If not, create and LoadOrStore a new one.
+			client = s.newClient(clientURL)
+			s.clients.Store(clientURL.String(), client)
+		}
+		client.count++
```

Comment:
Why are we blocking globally when we could do something fairly straightforward and non-blocking like:

```suggestion
		for {
			// Create a client
			client = s.newClient(clientURL)
			atomic.AddInt64(&client.count,1)
			var loaded bool
			// Load or store the client
			client, loaded = s.clients.LoadOrStore(clientURL.String(),client)
			// If its loaded and incrementing the counter from 0... somebodies already deleted it, so try again
			if loaded && atomic.AddInt64(&client.count,1) > 1 {
				continue
			}
		}
```

Reply:
Ah... I'd missed that this was the multiexecutor and thus per clientURL and not global.  This is fine as written.